### ISSUE TITLE
Clang++ 9.0 and boost::optional rebind semantics in CUDA code

### ISSUE THREAD

**Issue by jngrad** [espressomd/espresso#3650]
I get a compiler error on `src/core/grid_based_algorithms/lbgpu_cuda.cu` with Clang++-9.0, boost 1.71 and CUDA 10.1 from `nvidia-cuda-toolkit` on Ubuntu 20.04. The two offending functions are:
https://github.com/espressomd/espresso/blob/48ba8d16152ca44e09cb2da2a03b2e4d99665a4b/src/core/grid_based_algorithms/lbgpu_cuda.cu#L3042-L3048

According to the [boost::optional docs](https://www.boost.org/doc/libs/1_71_0/libs/optional/doc/html/boost_optional/rebinding_semantics_for_assignment_of_optional_references.html), we should be able to rebind an optional using the assignment operator, even if that optional is already initialized. The following MWE compiles just fine as a .cpp file, but fails as a .cu file:
```c++
#include <boost/optional.hpp>
template <typename T>
class A {
public:
  A(T &a) : x(a) {}
  T x;
};
boost::optional<A<int>> rng_counter_coupling_gpu;

void foo(int a) {
  rng_counter_coupling_gpu = A<int>(a);
}
```
```
$> clang++-9 -pedantic -fPIC -std=c++14 -c mwe.cpp
$> clang++-9 -pedantic -fPIC -std=c++14 --cuda-gpu-arch=sm_52 -c mwe.cu
mwe.cu:11:28: error: use of overloaded operator '=' is ambiguous (with operand types 'boost::optional<A<int> >' and 'A<int>')
  rng_counter_coupling_gpu = A<int>(a);
  ~~~~~~~~~~~~~~~~~~~~~~~~ ^ ~~~~~~~~~
/usr/include/boost/optional/optional.hpp:993:5: note: candidate function [with Expr = A<int>]
    operator= ( Expr&& expr )
    ^
/usr/include/boost/optional/optional.hpp:1064:5: note: candidate function [with T_ = A<int>]
    operator= ( T_&& val )
    ^
/usr/include/boost/optional/optional.hpp:1047:15: note: candidate function
    optional& operator= ( optional && ) = default;
              ^
/usr/include/boost/optional/optional.hpp:1035:15: note: candidate function
    optional& operator= ( optional const& rhs ) = default;
              ^
1 error generated when compiling for sm_52.
```
It's not reproducible on the ICP machines with `clang++-9 -pedantic -fPIC -std=c++14 --cuda-gpu-arch=sm_52 --cuda-path=/usr/local/cuda-10.0 -c mwe.cu`. I had to start `docker run --user espresso --runtime=nvidia espressomd/docker-ubuntu-20.04:01cf23487c9eeec758daf425c9947d8089ed9da0 bash` and install `nvidia-cuda-toolkit`.

Making `foo()` check the initialization status of the optional doesn't help:
```c++
void foo(int a) {
  if(rng_counter_coupling_gpu) {
    *rng_counter_coupling_gpu = A<int>(a);
  } else {
    rng_counter_coupling_gpu = A<int>(a);
  }
}
```
Making `foo()` always derefence is wrong, but compilation of the .cu file finally succeeds:
```c++
void foo(int a) {
  *rng_counter_coupling_gpu = A<int>(a);
}
```
It's wrong because dereferencing an un-initialized boost optional will cause an assertion to throw, and this is indeed what happens if you do that in a .cpp file with `int main() { foo(1); }`. Yet ESPResSo compiles with this incorrect workaround in `lbgpu_cuda.cu`, and all the tests pass.

What am I missing here?

**Comment by fweik** [comment]
You can explicitly pass an `optional` to make it unambiguous? Like
```c++
rng_counter_coupling_gpu = boost::make_optional<A>(a);
```
I'm not sure if I follow your MWE: I don't think it's relevant that the type is a template instantiation,
and it's my understanding that the ambiguous operator is on `boost::optional` and not on the contained type.  ~~It's btw. a  known problem pattern that unconstrained template-parameterized forwarding reference special functions hide or collide with copy and move special functions (also for constructors). From the shown code I can't tell why this only happens in cuda code, I'm sure there are some special rules for cuda, because why not nvidia?!~~ Seems to be an actual problem with the boost compiler detection...

**Comment by fweik** [comment]
This is a version of https://github.com/boostorg/config/issues/297 no?

**Comment by jngrad** [comment]
> This is a version of [boostorg/config#297](https://github.com/boostorg/config/issues/297) no?

Indeed, I missed that post. I'll try that tomorrow.

**Comment by jngrad** [comment]
Patching the boost library as explained in boostorg/config#297 fixed the compiler error, yet introduced a new compiler error for structs, classes and templated classes, but not for basic types like `int`:
```
In file included from mwe.cu:1:
In file included from /usr/include/boost/optional.hpp:15:
In file included from /usr/include/boost/optional/optional.hpp:43:
In file included from /usr/include/boost/type_traits/is_base_of.hpp:12:
/usr/include/boost/type_traits/is_base_and_derived.hpp:142:25: error: invalid application of 'sizeof' to an incomplete type 'boost::in_place_factory_base'
    BOOST_STATIC_ASSERT(sizeof(B) != 0);
                        ^~~~~~~~~
/usr/include/boost/static_assert.hpp:70:55: note: expanded from macro 'BOOST_STATIC_ASSERT'
#     define BOOST_STATIC_ASSERT( ... ) static_assert(__VA_ARGS__, #__VA_ARGS__)
                                                      ^~~~~~~~~~~
/usr/include/boost/type_traits/is_base_and_derived.hpp:217:41: note: in instantiation of template class 'boost::detail::is_base_and_derived_impl2<boost::in_place_factory_base, A<int> >' requested here
    BOOST_STATIC_CONSTANT(bool, value = bound_type::value);
                                        ^
/usr/include/boost/type_traits/is_base_of.hpp:25:31: note: in instantiation of template class 'boost::detail::is_base_and_derived_impl<boost::in_place_factory_base, A<int> >' requested here
            (::boost::detail::is_base_and_derived_impl<ncvB,ncvD>::value) ||
                              ^
/usr/include/boost/type_traits/is_base_of.hpp:31:58: note: in instantiation of template class 'boost::detail::is_base_of_imp<boost::in_place_factory_base, A<int> >' requested here
      : public integral_constant<bool, (::boost::detail::is_base_of_imp<Base, Derived>::value)> {};
                                                         ^
/usr/include/boost/optional/optional.hpp:794:32: note: in instantiation of template class 'boost::is_base_of<boost::in_place_factory_base, A<int> >' requested here
  : boost::conditional< boost::is_base_of<boost::in_place_factory_base, BOOST_DEDUCED_TYPENAME boost::decay<U>::type>::value
                               ^
/usr/include/boost/optional/optional.hpp:818:59: note: in instantiation of template class 'boost::optional_detail::is_convertible_to_T_or_factory<A<int>, A<int> >' requested here
  : boost::conditional< !is_optional_related<U>::value && is_convertible_to_T_or_factory<T, U>::value
                                                          ^
/usr/include/boost/core/enable_if.hpp:41:41: note: in instantiation of template class 'boost::optional_detail::is_optional_val_init_candidate<A<int>, A<int> >' requested here
  struct enable_if : public enable_if_c<Cond::value, T> {};
                                        ^
/usr/include/boost/optional/optional.hpp:992:35: note: in instantiation of template class 'boost::enable_if<boost::optional_detail::is_optional_val_init_candidate<A<int>, A<int> >, boost::optional<A<int> > &>' requested here
    BOOST_DEDUCED_TYPENAME boost::enable_if<optional_detail::is_optional_val_init_candidate<T, Expr>, optional&>::type
                                  ^
mwe.cu:11:28: note: while substituting deduced template arguments into function template 'operator=' [with Expr = A<int>]
  rng_counter_coupling_gpu = A<int>(a);
                           ^
/usr/include/boost/optional/detail/optional_factory_support.hpp:32:9: note: forward declaration of 'boost::in_place_factory_base'
  class in_place_factory_base ;
        ^
In file included from mwe.cu:1:
In file included from /usr/include/boost/optional.hpp:15:
In file included from /usr/include/boost/optional/optional.hpp:43:
In file included from /usr/include/boost/type_traits/is_base_of.hpp:12:
/usr/include/boost/type_traits/is_base_and_derived.hpp:142:25: error: invalid application of 'sizeof' to an incomplete type 'boost::typed_in_place_factory_base'
    BOOST_STATIC_ASSERT(sizeof(B) != 0);
                        ^~~~~~~~~
/usr/include/boost/static_assert.hpp:70:55: note: expanded from macro 'BOOST_STATIC_ASSERT'
#     define BOOST_STATIC_ASSERT( ... ) static_assert(__VA_ARGS__, #__VA_ARGS__)
                                                      ^~~~~~~~~~~
/usr/include/boost/type_traits/is_base_and_derived.hpp:217:41: note: in instantiation of template class 'boost::detail::is_base_and_derived_impl2<boost::typed_in_place_factory_base, A<int> >' requested here
    BOOST_STATIC_CONSTANT(bool, value = bound_type::value);
                                        ^
/usr/include/boost/type_traits/is_base_of.hpp:25:31: note: in instantiation of template class 'boost::detail::is_base_and_derived_impl<boost::typed_in_place_factory_base, A<int> >' requested here
            (::boost::detail::is_base_and_derived_impl<ncvB,ncvD>::value) ||
                              ^
/usr/include/boost/type_traits/is_base_of.hpp:31:58: note: in instantiation of template class 'boost::detail::is_base_of_imp<boost::typed_in_place_factory_base, A<int> >' requested here
      : public integral_constant<bool, (::boost::detail::is_base_of_imp<Base, Derived>::value)> {};
                                                         ^
/usr/include/boost/optional/optional.hpp:795:32: note: in instantiation of template class 'boost::is_base_of<boost::typed_in_place_factory_base, A<int> >' requested here
                     || boost::is_base_of<boost::typed_in_place_factory_base, BOOST_DEDUCED_TYPENAME boost::decay<U>::type>::value
                               ^
/usr/include/boost/optional/optional.hpp:818:59: note: in instantiation of template class 'boost::optional_detail::is_convertible_to_T_or_factory<A<int>, A<int> >' requested here
  : boost::conditional< !is_optional_related<U>::value && is_convertible_to_T_or_factory<T, U>::value
                                                          ^
/usr/include/boost/core/enable_if.hpp:41:41: note: in instantiation of template class 'boost::optional_detail::is_optional_val_init_candidate<A<int>, A<int> >' requested here
  struct enable_if : public enable_if_c<Cond::value, T> {};
                                        ^
/usr/include/boost/optional/optional.hpp:992:35: note: in instantiation of template class 'boost::enable_if<boost::optional_detail::is_optional_val_init_candidate<A<int>, A<int> >, boost::optional<A<int> > &>' requested here
    BOOST_DEDUCED_TYPENAME boost::enable_if<optional_detail::is_optional_val_init_candidate<T, Expr>, optional&>::type
                                  ^
mwe.cu:11:28: note: while substituting deduced template arguments into function template 'operator=' [with Expr = A<int>]
  rng_counter_coupling_gpu = A<int>(a);
                           ^
/usr/include/boost/optional/detail/optional_factory_support.hpp:33:9: note: forward declaration of 'boost::typed_in_place_factory_base'
  class typed_in_place_factory_base ;
        ^
2 errors generated when compiling for sm_52.
```

**Comment by RudolfWeeber** [comment]
this is also present when building with clang 8.

How big of an issue would it be to switch the dev branch to C++17? Then, we could use std::optional (after checking that that actually works, that is)

What to do about 4.1 is a different issue, of course. I did not check, if the counters are already optionals, there.
If the counters are the only instance, we can probably work around that issue for 4.1.

**Comment by fweik** [comment]
So, what version of NVCC supports C++17? (The alternative would be abseil, which provides C++14 implementations of the C++17 library parts). Also `boost::optional` and `std::optional` are not exactly equivalent (`std::optional<T&>` is not a thing, but does work with boost, don't know if we have that).

**Comment by jngrad** [comment]
offline core team discussion: remove the optional from the CUDA code

**Comment by jngrad** [comment]
I've replaced the `boost::optional` by a custom implementation in f18fc3b359360e. It's now possible to compile with Clang 9, but Clang-Tidy fails on .cu files. We need to either figure out why, or find a way in CMake to exclude CUDA files from Clang-Tidy. MWE:
```
$> export ASAN_OPTIONS="allocator_may_return_null=1" OMPI_CC=clang-9 OMPI_CXX=clang++-9 \
     CC=clang-9 CXX=clang++-9 UBSAN_OPTIONS=suppressions=/home/espresso/espresso/maintainer/CI/ubsan.supp
$> cp ../maintainer/configs/maxset.hpp myconfig.hpp
$> cmake .. -DWITH_CCACHE=ON -DWITH_CUDA=ON -DWITH_UBSAN=ON -DWITH_ASAN=ON \
     -DWITH_CLANG_TIDY=ON -DCMAKE_BUILD_TYPE=Debug -DCUDA_NVCC_EXECUTABLE=$(which clang++-9)
[  7%] Building CXX object src/core/CMakeFiles/EspressoCore.dir/actor/DipolarBarnesHut_cuda.cu.o
error: unable to handle compilation, expected exactly one compiler job in ' "/usr/lib/llvm-9/bin/clang-tool" "-cc1" [...] "-x" "cuda" "/home/espresso/t/espresso/src/core/actor/DipolarBarnesHut_cuda.cu"; ' [clang-diagnostic-error]
Error while processing /home/espresso/t/espresso/src/core/actor/DipolarBarnesHut_cuda.cu.
Found compiler error(s).
```
The message was truncated. The complete trace is attached: [clang-tidy-trace.log](https://github.com/espressomd/espresso/files/4495684/clang-tidy-trace.log). I took the liberty of re-wrapping the trace, as one line was 11'000 characters long.

I looked into Clang 8 using Rudolf's dockerfiles, but `/usr/bin/clang++-8 --cuda-path=/usr/lib/cuda -v` returns `Found CUDA installation: /usr/lib/cuda, version unknown`, which suggests it's incompatible with CUDA 10.1. If I try to build anyway, the following compiler error is raised:
```
$> export ASAN_OPTIONS="allocator_may_return_null=1" OMPI_CC=clang-8 OMPI_CXX=clang++-8 \
     CC=clang-8 CXX=clang++-8 UBSAN_OPTIONS=suppressions=/home/espresso/espresso/maintainer/CI/ubsan.supp
$> cp ../maintainer/configs/maxset.hpp myconfig.hpp
$> cmake .. -DWITH_CCACHE=ON -DWITH_CUDA=ON -DWITH_UBSAN=ON -DWITH_ASAN=ON \
     -DWITH_CLANG_TIDY=ON -DCMAKE_BUILD_TYPE=Debug -DCUDA_NVCC_EXECUTABLE=$(which clang++-8) \
     -DCMAKE_CXX_FLAGS=--cuda-path=/usr/lib/cuda -DCUDA_NVCC_FLAGS=--cuda-path=/usr/lib/cuda
[  5%] Building CXX object src/core/CMakeFiles/EspressoCore.dir/actor/DipolarBarnesHut_cuda.cu.o
error: cannot find libdevice for sm_30. Provide path to different CUDA installation via --cuda-path, or pass -nocudalib to build without linking with libdevice. [clang-diagnostic-error]
error: cannot find libdevice for sm_52. Provide path to different CUDA installation via --cuda-path, or pass -nocudalib to build without linking with libdevice. [clang-diagnostic-error]
error: unable to handle compilation, expected exactly one compiler job in ' "/usr/lib/llvm-8/bin/clang-tool" "-cc1" "-triple" [...] "-x" "cuda" "/home/espresso/t/espresso/src/core/actor/DipolarBarnesHut_cuda.cu"; ' [clang-diagnostic-error]
Error while processing /home/espresso/t/espresso/src/core/actor/DipolarBarnesHut_cuda.cu.
Found compiler error(s).
```
It's the same Clang-Tidy error, plus an extra `error: cannot find libdevice` that comes from Clang. Without Clang-Tidy:
```
[ 20%] Building CXX object src/core/CMakeFiles/EspressoCore.dir/actor/DipolarBarnesHut_cuda.cu.o
clang: error: cannot find libdevice for sm_30. Provide path to different CUDA installation via --cuda-path, or pass -nocudalib to build without linking with libdevice.
clang: error: cannot find libdevice for sm_52. Provide path to different CUDA installation via --cuda-path, or pass -nocudalib to build without linking with libdevice.
```
Same outcome for Clang 7.

**Comment by mkuron** [comment]
> `Found CUDA installation: /usr/lib/cuda`, version unknown, which suggests it's incompatible with CUDA 10.1.

Clang has a list of CUDA versions whose libraries and nvptx assembler it supports. As per https://github.com/llvm/llvm-project/blame/a9627b7ea7e2b47488188cb1d80a23b24a8066b2/clang/include/clang/Basic/Cuda.h#L25, CUDA 9.0 is supported by Clang 6.0+, CUDA 9.1 and 9.2 by Clang 7.0+, CUDA 10.0 by Clang 8.0+, CUDA 10.1 by Clang 9.0+, and CUDA 10.2 and 11.0 only by the current Clang master.

> find a way in CMake to exclude CUDA files from Clang-Tidy

That would mean not setting `CUDA_NVCC_EXECUTABLE=$(which clang++)`, i.e. not compiling CUDA code with Clang instead of nvcc. I strongly advice against that as Clang's warnings and static analysis provide so much more useful diagnostics than nvcc (nvcc has next to no diagnostics except for parse errors for the GPU code, and only allows the host compiler to give very limited diagnostics on the host code).

**Comment by fweik** [comment]
@jngrad ~~this may be because the device code versions `sm_30` and `sm_52` are no longer supported by the cuda installation?~~ according to wikipedia this is supported

**Comment by mkuron** [comment]
> `unable to handle compilation, expected exactly one compiler job in ' "/usr/lib/llvm-9/bin/clang-tool" "-cc1"`

That's an odd message. Sounds like Clang is getting invoked with wrong arguments.

This looks like it explains what is happening: http://lists.llvm.org/pipermail/cfe-dev/2015-December/046634.html. Does that `clang-tool` invocation come from CMake? It probably needs to be invoked twice, once for host code and once for device code to do clang-tidy (though not to compile the code). Maybe the old CMake used to do just that, or it passed some flag to `clang-tidy` to check both host and device code, or the old Clang version automatically checked both.

**Comment by jngrad** [comment]
> Does that `clang-tool` invocation come from CMake?

Not sure, with the `clang:6.0` image the Clang-Tidy command for .cu files is completely different.

> To make this work you'd need to teach clang-check to handle this case
> properly, or you'd need to convince the clang driver to invoke the
> compiler only once (for just the host, or just the device).  The
> relevant flags to clang proper are --cuda-device-only or
> --cuda-host-only; I don't know offhand how to make those interact with
> clang-check.

I can tinker with CMake to see how Clang-Tidy is invoked, or choose whether to analyse only the host or device code with these flags.

**Comment by jngrad** [comment]
> Does that `clang-tool` invocation come from CMake?

CMake injects ` /usr/bin/cmake -E __run_co_compile --tidy="/usr/bin/clang-tidy-6.0;--line-filter=[{name: boost/mpi/collectives/all_reduce.hpp}]" --source=/work/jgrad/espresso/src/core/actor/DipolarBarnesHut_cuda.cu --` before the compilation command `/usr/bin/clang++ ...`. The feature `__run_co_compile` is undocumented, and the source code ([Kitware/CMake:Source/cmcmd.cxx#L406](https://github.com/Kitware/CMake/blob/df2d39bc512bd09ae808e8a2a6c7523737bb26c0/Source/cmcmd.cxx#L406)) doesn't show directly how the Clang-Tidy command is constructed.

The program `/usr/lib/llvm-9/bin/clang-tool` doesn't actually exist. Is it a placeholder?

> I strongly advice against that as Clang's warnings and static analysis provide so much more useful diagnostics than nvcc

Do you have an example of a C++ error that can only trigger a Clang-Tidy warning (by that I mean not a Clang compiler warning that Clang-Tidy forwards with a tag), and that will be detected from inside a .cu file? For example, the following piece of code raises multiple `clang-tidy-6.0` warnings in a .cpp file, but zero warning inside a .cu file, no matter whether I add `__host__`, `__device__`, both or none to the functions. I've tinkered for a few hours with Clang-Tidy on .cu files, without success.

```c++
// ill-formed functions to trigger Clang-Tidy warnings
char clang_tidy_1(char *p) {  // [readability-non-const-parameter]
  return *p;
}
int clang_tidy_2(int b) {
  if (b) return 1;
  else return 2;  // [readability-else-after-return]
}
double clang_tidy_3() {
  double d;
  int i = 42;
  d = 32 * 8 / (2 + i);  // [bugprone-integer-division]
  d = i / (1 << 4);  // [bugprone-integer-division]
  return d;
}
```

Adding `--cuda-device-only` to the CMake command that calls Clang-Tidy 9.0 doesn't change the error message, while `--cuda-host-only` fixes it, but then are we missing all the device code? I cannot check it at the moment, because the CUDA driver on my ICP machine is currently broken.

**Comment by fweik** [comment]
I doubt that `clang-tidy` can check device code, is there any documentation to that effect?

**Comment by mkuron** [comment]
It definitely checked host code when I added clang-tidy a few years ago. Does it do that when you add `--cuda-host-only`? I think it used to do device code too, but your sample code might not trigger it because almost any function is inlined by default — that is, your code is never compiled.

**Comment by mkuron** [comment]
```cuda
#include <cstdlib>

__device__ double clang_tidy_3() {
  double d;
  int i = 42;
  d = 32 * 8 / (2 + i);
  d = i / (1 << 4);
  return d;
}

__global__ void test(double * arr) {
    arr[0] = clang_tidy_3();
}

int main() {
	int * a = (int*) malloc(sizeof(int));
	
	if(a)
		return 1;
	else
		return 0;
}
```

`clang-tidy-9 --extra-arg=--cuda-host-only  test.cu` and `clang-tidy-9 --extra-arg=--cuda-device-only  test.cu` both result in
```
/tmp/test.cu:6:3: warning: Value stored to 'd' is never read [clang-analyzer-deadcode.DeadStores]
  d = 32 * 8 / (2 + i);
  ^
/tmp/test.cu:6:3: note: Value stored to 'd' is never read
/tmp/test.cu:19:10: warning: Potential leak of memory pointed to by 'a' [clang-analyzer-unix.Malloc]
                return 1;
                       ^
/tmp/test.cu:16:19: note: Memory is allocated
        int * a = (int*) malloc(sizeof(int));
                         ^
/tmp/test.cu:18:5: note: Assuming 'a' is non-null
        if(a)
           ^
/tmp/test.cu:18:2: note: Taking true branch
        if(a)
        ^
/tmp/test.cu:19:10: note: Potential leak of memory pointed to by 'a'
                return 1;
                       ^
```
suggesting that it always analyzes both host and device code.

**Comment by jngrad** [comment]
Thanks @mkuron! It's clear now that our CMake logic is incorrect. Adding `--cuda-host-only` to the list of compiler flags for CUDA files in docker image `clang:6.0` revealed 147 Clang-Tidy warnings:
```
make -j24 EspressoCore 2>&1 | grep 'treated as errors'
5 warnings treated as errors
13 warnings treated as errors
4 warnings treated as errors
16 warnings treated as errors
6 warnings treated as errors
40 warnings treated as errors
21 warnings treated as errors
12 warnings treated as errors
6 warnings treated as errors
24 warnings treated as errors
```
The majority are trivial to fix, but there are a few in there about potential memory leaks with malloc. Device code is also checked with `--cuda-host-only`.

We cannot simply add this flag to the compiler flags, since the generated shared objects lack device code: `CUDA error: invalid device symbol`. We also cannot do two compilations, once with `--cuda-host-only` and once with `--cuda-device-only`, because the latter throws `clang: error: cannot specify -o when generating multiple output files` with Clang-Tidy.

**Comment by mkuron** [comment]
> We cannot simply add this flag to the compiler flags

We would want to add `--extra-arg=--cuda-host-only test.cu` to the commands that clang-tidy is invoked with. Clang-tidy and the compilation itself are separate processes and it should be possible to pass arguments to one but not the other. Maybe directly add it here: https://github.com/espressomd/espresso/blob/python/CMakeLists.txt#L378 ? That's what they do in https://cmake.org/cmake/help/latest/variable/CMAKE_LANG_CLANG_TIDY.html.

**Comment by RudolfWeeber** [comment]
Se Ubuntu 20.04 Pr for clang tidy args.

**Comment by jngrad** [comment]
The following change restores the expected behavior of Clang-Tidy, but only for Clang-Tidy 6:
```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 86b96f7..d44d434 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -379,3 +379,3 @@ if(WITH_CLANG_TIDY)
   set(CMAKE_CXX_CLANG_TIDY
-      "${CLANG_TIDY_EXE};--line-filter=${CLANG_TIDY_EXCLUDES}")
+      "${CLANG_TIDY_EXE};--line-filter=${CLANG_TIDY_EXCLUDES};--extra-arg=--cuda-host-only")
 endif()
```
With that change, Clang-Tidy 9 doesn't fail anymore, but doesn't output any warning in the Ubuntu 20.04 image with `nvidia-cuda-toolkit`. Also, in that image, I can't run GPU scripts:
```
Cuda Memory error at /home/espresso/espresso/src/core/grid_based_algorithms/lbgpu_cuda.cu:2844.
CUDA error: no CUDA-capable device is detected
--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 0 in communicator MPI_COMMUNICATOR 4
with errorcode 1.
```
The GPUs appear in the `lspci` output, but the subfolders in `/usr/lib/cuda` are mostly empty, `nvidia-smi` is not even installed. I cannot install any of the `nvidia-driver-4xx` driver due to conflicts with the `nvidia-cuda-toolkit` package.

**Comment by mkuron** [comment]
> Also, in that image, I can't run GPU scripts

That image needs to be run using `nvidia-docker`. However, because it is not derived from Nvidia's official images, you need to add the appropriate metadata yourself, see https://github.com/NVIDIA/nvidia-docker/wiki/Usage#dockerfiles and https://gitlab.com/nvidia/container-images/cuda/-/blob/master/ubuntu/base/Dockerfile.jinja#L75.

> With that change, Clang-Tidy 9 doesn't fail anymore, but doesn't output any warning 

That's odd. `--extra-arg=--cuda-host-only` works locally for me as posted [above](https://github.com/espressomd/espresso/issues/3650#issuecomment-616059087).

**Comment by jngrad** [comment]
Yes, the stand-alone command `clang-tidy-9 --extra-arg=--cuda-host-only test.cu` works, but not through our CMake logic.

**Comment by jngrad** [comment]
I ended up using `strace -o trace.txt -s 80 -f make EspressoCore` to dump the complete trace of system calls and reveal the full command line for Clang-Tidy 9. Mystery solved: the culprit was option `"--line-filter=[{name: boost/mpi/collectives/all_reduce.hpp}]"`, which actually suppresses all warnings in source files (both .cpp and .cu). Here is a MWE:
```
$> clang-tidy-9 --line-filter='[{"name":"qwerty"}]' --extra-arg=--cuda-host-only test.cu
143 warnings generated.
Suppressed 143 warnings (138 in non-user code, 5 due to line filter).
Use -header-filter=.* to display errors from all non-system headers.
Use -system-headers to display errors from system headers as well.
```

**Comment by jngrad** [comment]
Found it: the `--line-filter` argument is actually inclusive, not exclusive. The correct syntax is therefore `--line-filter="[{'name': 'boost/mpi/collectives/all_reduce.hpp', 'lines': [[9999999,9999999]]}, {'name': '.h'}, {'name': '.hpp'}, {'name': '.hxx'}, {'name': '.hh'}, {'name': '.c'}, {'name': '.cpp'}, {'name': '.cxx'}, {'name': '.cc'}, {'name': '.cu'}, {'name': '.cuh'}]"`.

**Comment by RudolfWeeber** [comment]
>Found it: the --line-filter argument is actually inclusive, not exclusive. The correct syntax is therefore --line-filter="[{'name': >'boost/mpi/collectives/all_reduce.hpp', 'lines': [[9999999,9999999]]}, {'name': '.h'}, {'name': '.hpp'}, {'name': >'.hxx'}, {'name': '.hh'}, {'name': '.c'}, {'name': '.cpp'}, {'name': '.cxx'}, {'name': '.cc'}, {'name': '.cu'}, {'name': >'.cuh'}]".

Thanks. I apparently misunderstood how that works.

**Comment by jngrad** [comment]
We're not the only ones who got bitten by this mechanism. There is a PR from 2016 to add a flag `-suppress-checks-filter` to exclude files (https://reviews.llvm.org/D26418), but it hasn't made it yet.

**Comment by jngrad** [comment]
In the end, the `--line-filter` was useless, it can't suppress the boost warning:
```
$> clang-tidy-9 \
     --line-filter="[{'name':'src/core/cells.cpp'}]" \ # exclude all but the current file
     /work/jgrad/espresso/src/core/cells.cpp -- /usr/bin/clang++-9 ...
1 warning generated.
/usr/include/boost/mpi/collectives/all_reduce.hpp:59:3: error: Address of
stack memory associated with local variable 'op' is still referred to by the
global variable 'op_ptr' upon returning to the caller.  This will be a dangling
reference [clang-analyzer-core.StackAddrEscapeBase,-warnings-as-errors]
```
I guess that's not suppressed because [`clang-analyzer-core.StackAddrEscapeBase`](https://clang.llvm.org/extra/clang-tidy/checks/clang-analyzer-core.StackAddressEscape.html) is not a Clang-Tidy rule, but an alias to [`core.StackAddressEscape`](https://clang.llvm.org/docs/analyzer/checkers.html#core-stackaddressescape-c). That would also explain why we can't disable it with `Checks: | '-clang-analyzer-core.StackAddrEscapeBase'`.

We can't exclude the boost MPI file with `HeaderFilterRegex` because it's inclusive instead of exclusive. The exclusive form was proposed in a 2017 PR that still hasn't been merged (https://reviews.llvm.org/D34654). We also can't write a regex that excludes files because their regex parser uses the [ERE syntax](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap09.html#tag_09_04), which doesn't support negative lookaheads (https://lists.llvm.org/pipermail/cfe-dev/2015-November/046203.html).

What we could do however is make this warning harmless with `WarningsAsErrors: '*,-clang-analyzer-core.StackAddrEscapeBase'`. It'll still show up in the CI log, but it won't interrupt the build.

**Comment by jngrad** [comment]
There are more warnings triggered by boost MPI, e.g. `clang-analyzer-optin.mpi.MPI-Checker`, which will need to be disabled in `WarningsAsErrors`. I'm addressing all Clang-Tidy 9 warnings, including in CUDA code, either by fixing them or adding `NOLINT` if the change is not trivial.

**Comment by mkuron** [comment]
> `clang-analyzer-optin.mpi.MPI-Checker`

That is pretty much incompatible with Boost.MPI because it doesn't properly trace MPI calls across multiple files. We have already disabled that via https://github.com/espressomd/espresso/blob/08dd1948402c60e82690b4525b3547fbaaf84938/.clang-tidy#L6

.

> /usr/include/boost/mpi/collectives/all_reduce.hpp:59

See https://github.com/boostorg/mpi/pull/52 and https://github.com/boostorg/mpi/issues/68. This issue should have been fixed since Boost 1.69. If not, please report again upstream.

**Comment by jngrad** [comment]
I was using the ICP machine with boost 1.65 for convenience, that must explain it.
